### PR TITLE
Add async client

### PR DESCRIPTION
**Pull request by nocarryr**
There is a bit of duplication from the thread-based client since there is no inheritance, but I was trying to avoid modifying existing code as much as possible.

The `aiohttp` dependency added is declared as an "extra" which could be installed with
```
pip install py-mjpeg[asyncio]
```
or
```
pip install -e ".[asyncio]"
```
for develop mode

### CODE REVIEW

File: mjpeg/__init__.py
Comment (janakj) [inline mjpeg/__init__.py]: Shouldn't you be awaiting aread_data() here instead read_data()?

File: mjpeg/__init__.py
Comment (janakj) [inline mjpeg/__init__.py]: If I read the above code correctly, you are trying to find a chunk_size that is a power of two and that is not bigger than length? If that is the case, shouldn't the boolean expression in the if condition read `chunk_size > length`?

The way you have it written now, the body of the if statement is only executed if `chunk_size <= length`, but then the while loop terminates immediately without any iterations because `chunk_size > length` will always evaluate to False.

File: mjpeg/__init__.py
Comment (janakj) [inline mjpeg/__init__.py]: asyncronously -> asynchronously

File: mjpeg/aioclient.py
Comment (janakj) [inline mjpeg/aioclient.py]: asyncronous -> asynchronous

File: mjpeg/aioclient.py
Comment (janakj) [inline mjpeg/aioclient.py:250]: Yes, I will merge the other PR, so please feel free to refactor.

**Review by janakj — CHANGES_REQUESTED** [review]
Thank you so much for contributing the async functionality! I would be happy to merge this in the current form (after you address my comments). I also wonder if we should try to reorganize the library somewhat to make the async functionality a first-class citizen and to address some of the code duplication. For example:
  * Would it make sense to move the async functions to a separate module to remove the need to prefix each asynchronous function name with "a"?
  * Should we try to build the synchronous functions on top of the asynchronous functions to eliminate the duplicated code?

In any case, we could try to address these later. For now, I would like to merge your PR to get the asynchronous functionality into the hands of people because it is very useful.

File: mjpeg/__init__.py
Comment (nocarryr) [reply mjpeg/__init__.py]: Wow, good catch!  Honestly, I have no idea why that hasn't prevented things from working in my tests tbh.

It may be that I haven't run into this code path in my tests, so I'll need to investigate after making that change

File: mjpeg/__init__.py
Comment (nocarryr) [reply mjpeg/__init__.py]: Another silly mistake and another good catch!

I'm honestly not incredibly satisfied with this logic to begin with.  It would be much cleaner if `aiohttp.BodyPartReader` had a buffer interface that would support `readinto()`.

I'm open to suggestions on making it cleaner and more efficient

File: mjpeg/__init__.py
Comment (nocarryr) [reply mjpeg/__init__.py]: I tested locally with the changes and obviously hadn't allowed the client to overrun before.

Working as expected now after fixing the `bytearray(range(left))` error in the line above.

File: mjpeg/aioclient.py
Comment (nocarryr) [reply mjpeg/aioclient.py:250]: I'd rather wait until merge if that's ok, just to keep each branch working on its own

File: mjpeg/__init__.py
Comment (janakj) [reply mjpeg/__init__.py]: Are you sure this is the correct fix? The way I read it now, if chunk_size is less than or equal to length, it will decrease chunk_size until it is one. Is that what you intended?

File: mjpeg/__init__.py
Comment (janakj) [inline mjpeg/__init__.py:78]: Is it necessary to create a byte array and copy data into it here? The function askip_data discards the byte array immediately. Would it be possible to simply read the data from the stream without copying it into another buffer?

MJPEG stream chunks can be quite big and I would like to try to minimize the overhead here.

**Comment by janakj** [comment]
Thanks a lot for the contribution! I have merged this PR for now. Before we release this to PyPI, I think the async functionality may benefit from another round of reviews and perhaps some testing.

A big chunk of the py-mjpeg library deals with the parsing of the HTTP-based streaming protocol. That was necessary because urllib did not have that functionality when I originally created this library. I was not aware of aiohttp until now, but it seems that it already has all the functionality one needs to read and parse the multi-part HTTP response. I wonder if it makes sense to model the response processing code in the asynchronous client after the synchronous variant. Perhaps there is a simpler way to implement this using aiohttp's APIs and classes.

**Comment by nocarryr** [comment]
Thanks for the merge, and I agree completely. There are areas I'd like to optimize and tweak.

I started working on methods of testing and can continue with building a test suite if that's something you'd be interested in.

As I was building this, I realized the same thing about the multi-part handling code. It would be worth looking into a more unified approach for both the async and sync code in the future, but that's better as a separate discussion I imagine.

**Comment by janakj** [comment]
> On Mar 18, 2023, at 17:26, nocarryr ***@***.***> wrote:
> Thanks for the merge, and I agree completely. There are areas I'd like to optimize and tweak.
> 
> I started working on methods of testing and can continue with building a test suite if that's something you'd be interested in.
> 
Absolutely!
> As I was building this, I realized the same thing about the multi-part handling code. It would be worth looking into a more unified approach for both the async and sync code in the future, but that's better as a separate discussion I imagine.
> 
Let me take a closer look at aiohttp. Then we can make a plan for a unified approach. I would not mind building the blocking API and client on top of the asynchronous code, for example. If that turns out to be difficult, we can always keep the two implementations separate.

-Jan

**Comment by nocarryr** [comment]
> Let me take a closer look at aiohttp. Then we can make a plan for a unified approach. I would not mind building the blocking API and client on top of the asynchronous code, for example. If that turns out to be difficult, we can always keep the two implementations separate.

It might be worth looking into [httpx](https://www.python-httpx.org/) as a solution for both sync and async.  It supports multipart, but at first glance it's hard to tell if multipart responses are available.

**Pull request merged**
